content/advent-2015/go-in-a-yocto-project.md

```diff
@@ -0,0 +1,147 @@
++++
+author = ["Marcelo Magallon"]
+date = "2015-12-13T00:00:00-06:00"
+linktitle = "Go in a Yocto project"
+series = ["Advent 2015"]
+title = "Integrating Go in a Yocto-based project"
++++
+
+From its website, [Yocto](https://www.yoctoproject.org/), part of the
+[Linux Foundation Collaborative
+Projects](http://collabprojects.linuxfoundation.org/), is <cite>an open
+source collaboration project that provides templates, tools and methods
+to help you create custom Linux-based systems for embedded products
+regardless of the hardware architecture</cite>. Given Go's [wonderful
+support for cross
+compilation](http://dave.cheney.net/2015/08/22/cross-compilation-with-go-1-5),
+the two are like a match made in heaven.
+
+While not a requirement, Yocto-based projects have a strong preference
+for building everything from source, including the toolchain. One of the
+biggest changes introduced with the 1.5 release was the move of the
+entire toolchain to Go, which poses an interesting bootstraping problem:
+you need a Go compiler to build a Go compiler. The developers solved
+this by introducing [a clear
+requirement](https://docs.google.com/document/d/1OaatvGhEAq7VseQ9kkavxKNAfepWy2yhPUBs96FGV28/edit)
+for the toolchain to be buildable, for as long as possible, using Go
+1.4.
+
+On the other hand, Yocto makes a distinction between programs meant to
+run on the host and programs meant to run on the target system. Yocto
+recipes can produce binaries that are required to run in the host during
+a later phase of the build process, called "native packages". A
+cross-compiler is a program that runs on the host system and produces
+binaries for the target ssytem.  Together with the bootstraping
+solution, this means in order to support Go, you need at least two
+packages:
+
+* A bootstrap Go 1.4 compiler (`go-bootstrap`), which is a native
+  package used to build the compiler for the desired version.
+* A Go 1.5 (or later) cross compiler (`go-cross`), which is a native
+  package used to build binaries for the target. This provides the
+  entire Go toolchain.
+
+There's a third type of program that's of interest: one that runs on the
+host but is not meant to produce binaries for the target. An example is
+a program that processes some input and generates output that's needed
+later in the build process, like `go generate`. To cover this case, a
```

Comment:
note: go generate "is intended to be used by the author of the Go package, not its clients." - [source](https://blog.golang.org/generate)

so it may not be the best example here. everything it generates should be checked in with the source code.